src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -434,17 +473,37 @@ public void EnsureSolutionIsLoaded()
             });
         }
 
-        public string SolutionDirectory => NuGetUIThreadHelper.JoinableTaskFactory.Run(GetSolutionDirectoryAsync);
+        public string SolutionDirectory
+        {
+            get
+            {
+                if (_solutionEventsCookie != 0)
+                {
+                    return _solutionDirectory;
+                }
+                else
+                {
+                    return NuGetUIThreadHelper.JoinableTaskFactory.Run(GetSolutionDirectoryAsync);
+                }
+            }
+        }
 
         public async Task<string> GetSolutionDirectoryAsync()
         {
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-            var vsSolution = await _asyncVSSolution.GetValueAsync();
-            if (IsSolutionOpenFromVSSolution(vsSolution))
+            if (_solutionEventsCookie != 0)
             {
-                return (string)GetVSSolutionProperty(vsSolution, (int)__VSPROPID.VSPROPID_SolutionDirectory);
+                return _solutionDirectory;
+            }
+            else
+            {
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                var vsSolution = await _asyncVSSolution.GetValueAsync();
+                if (IsSolutionOpenFromVSSolution(vsSolution))
+                {
+                    return (string)GetVSSolutionProperty(vsSolution, (int)__VSPROPID.VSPROPID_SolutionDirectory);
```

Comment:
Should we set `_solutionDirectory` variable value before returning the value to the caller? Given that we don't publish `NuGet.PackageManagement.VisualStudio` nupkg, I assume the callers for this API will be internal to NuGet. I am trying to understand in what scenarios callers will be invoking this method directly without invoking `InitializeAsync` method that caches all the required values such as solution open, solution directory etc.,

If we are planning to make any changes to address this comment, please consider making changes to `IsSolutionOpenAsync` method also.

Please correct me If I my understanding is incorrect. Thanks.

Reply:
no, because in the `else` block (so `_solutionEventsCookie == true`), `InitializeAsync` has not yet been called, and therefore we're not getting solution open/close events which we'd need to tell us when the solution directory changes.

There are a bunch of scenarios when that might happen, like opening Tools->Options, and going to any of the NuGet Package Manager options packages without opening a solution. Or, if some other component calls one of our MEF APIs before VsSolutionManager gets initialized (for example, a native C++ project).